zbuf/array.go

```diff
@@ -38,13 +54,27 @@ func (a *Array) Write(r *zed.Value) error {
 // or it returns nil if the Array is empty.
 func (a *Array) Read() (*zed.Value, error) {
 	var rec *zed.Value
-	if len(*a) > 0 {
-		rec = &(*a)[0]
-		*a = (*a)[1:]
+	if len(a.values) > 0 {
+		rec = &a.values[0]
+		a.values = a.values[1:]
 	}
 	return rec, nil
 }
 
 func (a Array) NewReader() zio.Reader {
 	return &a
 }
+
+func (a *Array) Context() expr.Context {
+	return a
+}
+
+func (*Array) NewValue(typ zed.Type, bytes zcode.Bytes) *zed.Value {
+	// XXX can make this more efficient later
+	return zed.NewValue(typ, bytes)
+}
+
+func (*Array) CopyValue(val zed.Value) *zed.Value {
+	// XXX can make this more efficient later
+	return zed.NewValue(val.Type, val.Bytes)
```

Comment:
If the CopyValue contract is that it returns a value that does not share val.Bytes, then we need an actual copy here.
```suggestion
	return val.Copy()
```

Reply:
This was consistent with current behavior.  I think we should fix it when we change the caller's to presume fate share between batch refcnt and the result from Eval().  Right now, they copy the value so no sense copying multiple times.  The other change needed is for the batch NewValue() alllocator to allocate zcode.Bytes inside of the batch memory pools.